src/components/Button/BistateButton.tsx

```diff
@@ -0,0 +1,76 @@
+/*! Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license. */
+import '../../yamui';
+import * as React from 'react';
+import Button, { ButtonProps, FabricButtonType } from './Button';
+
+export type CustomizableProps = 'color' | 'text' | 'ariaLabel' | 'icon' | 'iconPosition' | 'className';
+
+export type CustomizableButtonProps = Pick<ButtonProps, CustomizableProps>;
+
+export interface BistateButtonOwnProps {
+  activeProps: CustomizableButtonProps;
+}
+
+export type BistateButtonProps = BistateButtonOwnProps & ButtonProps;
+
+export interface BistateButtonState {
+  hasFocus: boolean;
+  isHovering: boolean;
+}
+
+const inactiveState = { hasFocus: false, isHovering: false };
+
+/**
+ * A `Button` that renders differently when 'active'; if the mouse is hovering or it has focus
+ */
+export default class BistateButton extends React.Component<BistateButtonProps, BistateButtonState> {
+  constructor(props: BistateButtonProps) {
+    super(props);
+    this.state = { ...inactiveState };
+  }
+
+  public render() {
+    const propsForState = this.isActive() ? this.props.activeProps : undefined;
+    const { children, activeProps, ...buttonProps } = this.props;
+    const props: ButtonProps = {
+      fullWidth: true,
```

Comment:
if you look at `<Button>`, @unindented implemented its loading state so the `text` is hidden from view/screenreaders but still rendered, ensuring the button width doesn't change when its text is replaced with a Spinner.

maybe `<Button>` could accept a prop like `longestAlternateText` which gets rendered in the same hidden manner to push the button to a max width? feels like an awkward API though... but we'd want to avoid setting a hard width on the button based on state because of perf, and also to allow the application to responsively change a button's width based on content or layout via `fullWidth: true`

Reply:
> maybe <Button> could accept a prop like longestAlternateText which gets rendered in the same hidden manner to push the button to a max width

How does a developer know what the 'longest' text is. Long is long in DOM terms - you'd have to stick it in the DOM to measure....... iii is less long than WWW

Reply:
in this case, BistateButton would pass either `text` or `activeProps.text` as Button's `text`, and pass the unused one as Button's `longestAlternateText`.

not suggesting this is a great API, just that the concept might work to push the button's width out to its max possible width without having the application try to hardcode widths. ideally BistateButton could handle this behind the scenes without surfacing a public `longestAlternateText` prop on Button...

Reply:
i don't follow. You can't know what the widest button will be (active or inactive) until its rendered to the dom and you measure

Reply:
I think PJ's naming is causing confusion here - I had the same issues initially.

I thin the suggestion is really for a 'secondarySizingText' property. Idea is that the button renders its regular content as normal, and then renders this text in some hidden way (clipped, transparent, etc). The longest of the two wins out as far as HTML's sizing goes.

Reply:
> Idea is that the button renders its regular content as normal, and then renders this text in some hidden way (clipped, transparent, etc). The longest of the two wins out as far as HTML's sizing goes

Sounds like putting it into the DOM to measure, then?

Reply:
Not really to measure (which suggests you'd do something with the results after); but just putting both strings in the DOM so that the button naturally gets the widest width of the two strings.

Net result is something like:

    <button>
        <div aria-hidden=true opacity=0 height=0>Inactive string</div>
        <div>Active String</div>
    </button>

(attrs to illustrate the concept only, apply salt as needed.)

Reply:
ick. ok

Reply:
Ick indeed! That's the sort of hackery I was referring to in my comment above. Didn't know we were already doing similar for the loading button though - so there's at least another legit use case for this sort of thing.

(Almost forgot - visibility:hidden is great for this sort of layout thing; it hides things visually and from a11y, but still does layout - so better than opacity hacking.)

Reply:
Here's the least hacky version I could come up with https://github.com/Microsoft/YamUI/pull/290/commits/c71b633cdd055376575ce445155ca02e26c3118d

Thoughts?

Reply:
(we might want to change children to widthPadder or some such, IDK. I dont have an informed opinion)